File: sda/lab1/BagIterator.cpp
#include "BagIterator.h"
#include "Bag.h"

BagIterator::BagIterator(const Bag& c) {
    this->c = c;
    this->idx = 0;
    this->how_many = 0;
    this->current = 0;
}

void BagIterator::first() {
    this->idx = 0;
    this->how_many = this->c.array[idx].apparitions;
    this->current = this->c.array[idx].element;
}

void BagIterator::next() {
    if (this->how_many > 0) {
        how_many -= 1;
    }
    else {
        this->idx += 1;
        if (this->valid()) {
            this->how_many = this->c.array[idx].apparitions;
            this->current = this->c.array[idx].element;
        }
    }
}

bool BagIterator::valid() const {
    return this->idx < c.size();
}

TElem BagIterator::getCurrent() const {
    return this->current;
}

File: sem2/dsa/lab2/BagIterator.py
class BagIterator:

    #creates an iterator for the Bag b, set to the first element of the bag, or invalid if the Bag is empty
    def __init__(self, b):
        self.bag = b
        self.idx = 0
        self.counter = 0

    # returns True if the iterator is valid
    # O(1)
    def valid(self):
        return self.idx < len(self.bag.elements)

    # returns the current element from the iterator.
    # throws ValueError if the iterator is not valid
    # O(1)
    def getCurrent(self):
        if not self.valid():
            raise ValueError
        return self.bag.elements[self.idx]

    # moves the iterator to the next element
    #throws ValueError if the iterator is not valid
    # O(1)
    def next(self):
        if not self.valid():
            raise ValueError
        self.counter += 1
        if self.counter == self.bag.occurrences[self.idx]:
            self.idx += 1
            self.counter = 0

    # sets the iterator to the first element from the Bag
    # O(1)
    def first(self):
        self.idx = 0
        self.counter = 0
